### ISSUE TITLE
Single quoted strings not supported?

### ISSUE THREAD

**Issue by binarycow** [PANTHEONtech/YANGinator#301]
Using a single quoted string results in errors

```yang
module alpha {
    namespace 'uri:alpha';
    prefix a;
    yang-version 1.1;
}
```

Errors produced:

- `<data def stmt>, <linkage stmts>, <meta stmts>, augment, deviation, extension, feature, grouping, identity, notification, revision, rpc or typedef expected, got '''`
- `<line break>, <uri str>, <wsp>, YangTokenType.DOUBLE_QUOTE, YangTokenType.FORWARD_SLASH or YangTokenType.SINGLE_QUOTE expected, got '''`

This works with no issues in pyang.  Additionally, the errors go away if I use a double quoted string for the namespace.

**Comment by binarycow** [comment]
Hey, thanks for this.

I don't know Java, but it seems every file you modified was related to URI parsing (i.e., the namespace statement, the one I mentioned in the ticket) 

I haven't tested to see if it works or not, but single quoted strings are allowed anywhere a double quoted string is allowed. Also, unquoted strings are allowed in those same spots.

And there's different post-processing for each type of string

- unquoted strings
  - no escapes
  - no concatenation
  - no trimming
  - single-line
  - cannot contain newlines, spaces, tabs, curly braces, semicolons, //, /*, or */
- single quoted strings
  - no escapes
  - supports concatenation (with single or double quoted strings) 
  - no trimming
  - single-line
  - cannot contain single quotes 
- double quoted strings
  - supports escapes
  - supports concatenation (with single or double quoted strings) 
  - automatic trimming (when multi-line) 
  - single-line or multi-line
  - can contain any character, if properly escaped

I havent checked if your tool supports all those scenarios - but all of those are supports for the argument of any statement (except input and output which don't take arguments)

**Comment by rovarga** [comment]
Right, so this is something where the RFC6020/RFC7950 grammar contradicts examples and real-world use (which is mostly derived from examples). The grammar does not allow for quotes, but then there is no guidance on URI escaping.

My guess is other parsers are not bitten by this because they deal with [lexical tokenization](https://www.rfc-editor.org/rfc/rfc7950#section-6.1) in their scanner, as (very weakly) hinted by the third paragraph [in section 14](https://www.rfc-editor.org/rfc/rfc7950#section-14):
`
   This grammar assumes that the scanner replaces YANG comments with a
   single space character.
`
meaning they effectively allow quoted content anywhere the ABNF mandates `< a string that matches the rule >`. `uri-str`, and others, should get the same treatment `path-str` gets -- for which we have `quoted-path-arg`.

**Comment by binarycow** [comment]
I don't think there is a contradiction. 


If you consider that:

- the RFC states "YANG modules are parsed as a series of tokens"
- the possible tokens are keyword, a string, a semicolon, or braces
- comments aren't considered a token, and are equivalent to whitespace (it's not in the list of token types, and the grammar specifies that it is considered to be equivalent to whitespace) 

then when the grammar says 

> uri-str = < a string that matches the rule > < URI in RFC 3986 >

I take that to mean "a string _token_ that matches the rule..." 

The RFC says that a string token can be a double quoted string, a single quoted string, or an unquoted string. 

So, I take that to mean that every statement (that accepts arguments) should be able to use all the features of strings. 

For example, this should be valid:

```
module 'foo' /* Hello */ + /* World! */ 'bar' {
  // remainder omitted for brevity
}
```

and should be equivalent to this:

```
module 'foobar' {
  // remainder omitted for brevity
}
```
